### PR TITLE
fix: fix the race condition when calling readAsDataURL after new Blob(blobs)

### PR DESCRIPTION
**Pull request by wood1986**
## Summary

```js
async () => {
  let blobs = [];
  for (let i = 0; i < 4; i++) {
    const res = await fetch();
    blobs = [...blobs, await res.blob()] 
  }
  const blob = new Blob(blobs); // <<<<<<<<<<<<<<< a
  return await new Promise((resolve, reject) => {
    const fileReader = new FileReader();
    fileReader.onload = async () => {
      await RNFS.writeFile(destPath, (fileReader.result as string).split(',')[1], 'base64');
      resolve(destPath);
    };
    fileReader.onabort = () => {
      reject('');
    };
    fileReader.onerror = (event) => {
      reject('');
    };
    fileReader.readAsDataURL(blob); // <<<<<<<<<<<<<<< b
  });
}
```

Sometime `fileReader.readAsDataURL` is unable to get blob from the dictionary after `new Blob(blobs)` and then reject with `Unable to resolve data for blob: blobId` in iOS or `The specified blob is invalid` in android. Because line `a` and `b` can be run in different thread. `new Blob([])` is in progress and `fileReader.readAsDataURL` accesses the blob dictionary ahead of the blob creation.

The expected behaviour is it should finish new Blob([]) first and then readAsDataURL(blob)

To fix that, there should be a lock inside the method `createFromParts`. For iOS, It needs to be a recursive_mutex to allow same thread to acquire lock

## Changelog

[iOS] [Fixed] - fix the race condition when calling readAsDataURL after new Blob(blobs)

### CODE REVIEW

**Comment by javache** [comment]
> Because line a and b can be run in different thread. 

All JS runs on a single thread in React Native. While your fix may be correct, this does not seem like the right root cause of the issue.

File: Libraries/Blob/RCTBlobManager.mm [hunk 1]
```diff
@@ -176,6 +176,7 @@ - (void)remove:(NSString *)blobId
 
 RCT_EXPORT_METHOD(createFromParts:(NSArray<NSDictionary<NSString *, id> *> *)parts withId:(NSString *)blobId)
 {
+  std::lock_guard<std::recursive_mutex> lock(_blobsMutex);
```
Comment (javache) [inline Libraries/Blob/RCTBlobManager.mm:179]: Why do we need a lock here? We only change instance state in `store` which already locks.

**Comment by wood1986** [comment]
Few things

> All JS runs on a single thread in React Native. While your fix may be correct, this does not seem like the right root cause of the issue.

Yes you are right. JS side is single thread. I was talking the native side

> Digging into the issue a bit more, I think I understand that on iOS there's no guarantee of ordering between RCTFileReaderModule and RCTBlobManager, which could cause a potential race. Using a recursive_mutex for this is not very elegant though, as there's not really any nested calls. Perhaps you could resolve this by changing RCTFileReaderModule to do in readAsDataURL: dispatch_async(blobManager.methodQueue, ^{ ... })

I fixed and it works

> On Android, all Native Modules run on a shared background thread, so I'm not sure what the issue is there.

Sorry for the confusion. I have not seen Android has the race condition. For the consistent reason, I made android same as iOS

**Review by javache — APPROVED** [review]

**Comment by wood1986** [comment]
Hey @javache is it possible to include this fix it in 0.69.2?

**Comment by javache** [comment]
> Hey @javache is it possible to include this fix in 0.69.2?

Please raise this in https://github.com/reactwg/react-native-releases/discussions/23

**Pull request closed without merging**